### ISSUE TITLE
timesyncd: NTP servers given by DHCP server are ignored

### ISSUE THREAD

**Issue by yuwata** [systemd/systemd#2850]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> systemd-229-6.fc25.x86_64

Note that systemd-222-14.fc23.x86_64 does not have the following problem.
### Used distribution

> fedora rawhide
### In case of bug report: Expected behaviour you didn't see

> systemd-timesyncd uses NTP servers given by DHCP server.
### In case of bug report: Unexpected behaviour you saw

> systemd-timesyncd ignores NTP servers given by DHCP server and uses fallback NTP servers.
### In case of bug report: Steps to reproduce the problem

> Configure .network with 'DHCP=yes' option.
> Start systemd-networkd.service and systemd-timesyncd.service.

I've tried 'UseNTP=true' option, which is the default, in [DHCP] section of .network file. However, the situation is the same.
Also, specifying both 'NTP=...' option in [Network] section and 'UseNTP=true' in [DHCP] section in .network file, then 'NTP=...' option is used.
Of course, only 'NTP=...' option is specified .network section, 'NTP=...' option is used.
When I list NTP servers in /etc/systemd/timesyncd.conf by 'NTP=...' option, then timesyncd uses the servers listed there.

**Comment by yuwata** [comment]
.network configuration file I use is the following.

```
[Match]
Name=localnet

[Network]
Description=Netowrk Settings for localnet
DHCP=yes
IPv6AcceptRouterAdvertisements=false

[DHCP]
UseHostname=no
UseDomains=true
CriticalConnection=yes
```

**Comment by yuwata** [comment]
```
# SYSTEMD_LOG_LEVEL=debug /usr/lib/systemd/systemd-timesyncd
Added new server 0.fedora.pool.ntp.org.
Added new server 1.fedora.pool.ntp.org.
Added new server 2.fedora.pool.ntp.org.
Added new server 3.fedora.pool.ntp.org.
systemd-timesyncd running as pid 4861
Selected server 0.fedora.pool.ntp.org.
Resolving 0.fedora.pool.ntp.org...
Resolved address 129.250.35.251:123 for 0.fedora.pool.ntp.org.
Resolved address 160.16.201.66:123 for 0.fedora.pool.ntp.org.
Resolved address 157.7.203.102:123 for 0.fedora.pool.ntp.org.
Resolved address 133.130.120.204:123 for 0.fedora.pool.ntp.org.
Selected address 129.250.35.251:123 of server 0.fedora.pool.ntp.org.
Connecting to time server 129.250.35.251:123 (0.fedora.pool.ntp.org).
Sent NTP request to 129.250.35.251:123 (0.fedora.pool.ntp.org).
NTP response:
  leap         : 0
  version      : 4
  mode         : 4
  stratum      : 2
  precision    : 0.000000 sec (-24)
  root distance: 0.113312 sec
  reference    : n/a
  origin       : 1458109811.988
  receive      : 1458109811.998
  transmit     : 1458109811.998
  dest         : 1458109812.002
  offset       : +0.003 sec
  delay        : +0.014 sec
  packet count : 1
  jitter       : 0.000
  poll interval: 64
  adjust (slew): +0.003 sec
  status       : 8193 sync
  time now     : 1458109812.004
  constant     : 2
  offset       : +0.003 sec
  freq offset  : +1680652 (25 ppm)
interval/delta/delay/jitter/drift 64s/+0.003s/0.014s/0.000s/+25ppm
Synchronized to time server 129.250.35.251:123 (0.fedora.pool.ntp.org).
```

**Comment by yuwata** [comment]
I found that if CriticalConnection=yes is set, then timesyncd does not use NTP servers given by a DHCP server. If CriticalConnection option is not set, then timesyncd correctly uses NTP servers given by the DHCP server. Is it expected behavior?

**Comment by poettering** [comment]
Sounds weird... CriticalConnection= should have no effect on NTP server data.

Do you see the NTP servers correctly in the "networkctl status -a" output?

**Comment by yuwata** [comment]
When CriticalConnection=yes is set,

```
# networkctl status
●        State: routable
       Address: 192.168.0.50 on localnet
                fe80::215:17ff:fe92:a0cc on localnet
       Gateway: 192.168.0.1 (Hewlett Packard) on localnet
                192.168.0.1 (Hewlett Packard) on localnet

# networkctl status -a
● 1: lo
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: n/a
            Type: loopback
           State: carrier (unmanaged)
             MTU: 65536
         Address: 127.0.0.1
                  ::1

● 2: localnet
       Link File: /etc/systemd/network/00-yitp00-localnet.link
    Network File: /etc/systemd/network/01-localnet.network
            Type: ether
           State: routable (configured)
            Path: pci-0000:01:00.0
          Driver: igb
          Vendor: Intel Corporation
      HW Address: 00:15:17:92:a0:cc (Intel Corporate)
             MTU: 1500
         Address: 192.168.0.50
                  fe80::215:17ff:fe92:a0cc
         Gateway: 192.168.0.1 (Hewlett Packard)
                  192.168.0.1 (Hewlett Packard)
             DNS: 192.168.0.1
  Search Domains: yukawa.kyoto-u.ac.jp
             NTP: 133.3.201.227

● 3: port2
       Link File: /etc/systemd/network/00-yitp00-port2.link
    Network File: n/a
            Type: ether
           State: off (unmanaged)
            Path: pci-0000:01:00.1
          Driver: igb
          Vendor: Intel Corporation
      HW Address: 00:15:17:92:a0:cd (Intel Corporate)
             MTU: 1500

# systemctl status systemd-timesyncd.service
● systemd-timesyncd.service - Network Time Synchronization
   Loaded: loaded (/usr/lib/systemd/system/systemd-timesyncd.service; disabled; vendor preset: disabled)
   Active: active (running) since Fri 2016-04-08 17:27:04 JST; 8min ago
     Docs: man:systemd-timesyncd.service(8)
 Main PID: 612 (systemd-timesyn)
   Status: "Synchronized to time server 202.181.103.212:123 (2.fedora.pool.ntp.org)."
    Tasks: 2 (limit: 512)
   CGroup: /system.slice/systemd-timesyncd.service
           └─612 /usr/lib/systemd/systemd-timesyncd

Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd[1]: Starting Network Time Synchronization...
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd[1]: Started Network Time Synchronization.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:04 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Network configuration changed, trying to establish connection.
Apr 08 17:27:34 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[612]: Synchronized to time server 202.181.103.212:123 (2.fedora.pool.ntp.org).
```

When CriticalConnection=yes is not set,

```
# networkctl status
●        State: routable
       Address: 192.168.0.50 on localnet
                fe80::215:17ff:fe92:a0cc on localnet
       Gateway: 192.168.0.1 (Hewlett Packard) on localnet
           DNS: 192.168.0.1
Search Domains: yukawa.kyoto-u.ac.jp
           NTP: 133.3.201.227

# networkctl status -a
● 1: lo
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: n/a
            Type: loopback
           State: carrier (unmanaged)
             MTU: 65536
         Address: 127.0.0.1
                  ::1

● 2: localnet
       Link File: /etc/systemd/network/00-yitp00-localnet.link
    Network File: /etc/systemd/network/01-localnet.network
            Type: ether
           State: routable (configured)
            Path: pci-0000:01:00.0
          Driver: igb
          Vendor: Intel Corporation
      HW Address: 00:15:17:92:a0:cc (Intel Corporate)
             MTU: 1500
         Address: 192.168.0.50
                  fe80::215:17ff:fe92:a0cc
         Gateway: 192.168.0.1 (Hewlett Packard)
             DNS: 192.168.0.1
  Search Domains: yukawa.kyoto-u.ac.jp
             NTP: 133.3.201.227

● 3: port2
       Link File: /etc/systemd/network/00-yitp00-port2.link
    Network File: n/a
            Type: ether
           State: off (unmanaged)
            Path: pci-0000:01:00.1
          Driver: igb
          Vendor: Intel Corporation
      HW Address: 00:15:17:92:a0:cd (Intel Corporate)
             MTU: 1500

# systemctl status systemd-timesyncd.service
● systemd-timesyncd.service - Network Time Synchronization
   Loaded: loaded (/usr/lib/systemd/system/systemd-timesyncd.service; disabled; vendor preset: disabled)
   Active: active (running) since Fri 2016-04-08 17:39:15 JST; 1min 29s ago
     Docs: man:systemd-timesyncd.service(8)
 Main PID: 608 (systemd-timesyn)
   Status: "Synchronized to time server 133.3.201.227:123 (133.3.201.227)."
    Tasks: 2 (limit: 512)
   CGroup: /system.slice/systemd-timesyncd.service
           └─608 /usr/lib/systemd/systemd-timesyncd

Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd[1]: Started Network Time Synchronization.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:15 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Network configuration changed, trying to establish connection.
Apr 08 17:39:45 yitp00.yukawa.kyoto-u.ac.jp systemd-timesyncd[608]: Synchronized to time server 133.3.201.227:123 (133.3.201.227).
```

**Comment by yuwata** [comment]
Why 'networkctl status' (without -a option) does not output DNS, search domains, and NTP when 'CriticalConnection=yes' is set?

**Comment by poettering** [comment]
I have no idea, not sure I follow what is going on. The CriticalConnection= setting should have no effect on any of these fields. Looks like a bug in networkd.

**Comment by martinpitt** [comment]
Fixed via PR #3080

**Comment by sebma** [comment]
> .network configuration file I use is the following.
> 
> ```
> [Match]
> Name=localnet
> 
> [Network]
> Description=Netowrk Settings for localnet
> DHCP=yes
> IPv6AcceptRouterAdvertisements=false
> 
> [DHCP]
> UseHostname=no
> UseDomains=true
> CriticalConnection=yes
> ```

@yuwata Hi, is the `.network` file your mentioning is located in the `/usr/lib/systemd/network/` directory  ?
